### PR TITLE
 [AIRFLOW-5264] Eslint + npmaudit

### PR DESCRIPTION
**Pull request by potiuk**
This is (initially experimental) change that introduces ESLiniting and npm audit as pre-commit hooks. Introduction of those two required fixing all current ESLint errors and upgrading NPM dependencies to latest versions (there were more than 50 vulnerabilities reported via npm audit with the current master version of the JS libraries).

It was initially experimental, because I was not sure whether this will be successful. I upgraded to latest version of dependencies using npm audit and then upgraded to latest version of eslint and finally run `eslint --fix`. Surprisingly there were just a few errors left that were easy to fix manually. Also when I run the webserver - going through pretty much all the screens I did not see any errors.

Of course it requires quite a bit more testing but I think if we make a small group effort to test it, we might get rid of all the ESLint problems and thanks to pre-commit hooks also keep it like that in the future. 

I am aware of Closed  #3656 and the idea from @mik-laj about separating out the HTML code to .js - and finally https://issues.apache.org/jira/browse/AIRFLOW-2803 which I created subtask of. However I think if we merge this one - this will be great step forward with the .js code improvement. Having eslint + npm audit in place is already great - especially from security point of view.

Note - I added npm audit as separate pre-commit hook, in the way that if someone changes any .js code, it will be run locally (and potentially point out to some upgrades that should be performed) but I skipped it in regular CI builds so that audit problems will not block us from merging unrelated changes. I run it however in the daily CRON build - so that we can get an early warning that some NPM dependency requires  upgrade due to security vulnerability.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5264

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by potiuk** [comment]
Hello @verdan @ashb @tedmiston @r39132 @jmcarp @feng-tao @mik-laj  - See the detailed description in the PR.

I know you were all involved in #3656 (which has been closed). But after I have done a little experiment, it turned out that eslinting for our .js files + upgrading to latest versions of deps (without known vulnerabilities) as first step can be introduced rather quickly. The change I've done in one evening already fixes all eslint problems AND upgrades to latest dependencies - this removes > 50 vulnerabilities from our code. Vast majority of those fixes is automatically generated by eslint. Even if we have js code to extract from html later - I think what I've done might be a major step in a very good direction.

I think if a few people will take a look and test it on their installation / test harnesses they have we might very quickly get this merged into master. I have not been too involved in web UI part so I might have made some basic mistakes and it's worth to take a deeper look (even thought the UI in general seems to work fine after the upgrades/linting).

I know for one there was a debate ES5 vs. ES6 in #3656 but it is likely not relevant any more (and especially in Airflow 2.0.0 context). So I went ahead and configured "es6" as target.

**Comment by ashb** [comment]
Do you know what the ES6 change means for the 1% (guessing) of Airflow users on IE? Do you know if as part of our compile asset stage we transpile the modules? (Sorry, about 7 yaks deep right now. I should just look myself)

I haven't looked at the code, but the description sounds good, and I'm I think I'm okay with the ES6 change - we can add a transpile step in our release/build pipeline later if we think it's worth while (and not already there)

**Comment by potiuk** [comment]
I am not 100% sure for the transpiling. It's a bit outside of my area (just a bit :) ). Maybe someone more familiar with the .js stack can chime in ? I see that we are using webpack and that we have babel-loader configured in webpack.config.js, but I am not sure if the configuration we have will actually produce an ES5-compliant javascript :). Anyone ?

**Comment by potiuk** [comment]
All is passing now :).

File: airflow/www/package.json [hunk 1]
```diff
@@ -57,16 +57,17 @@
   "dependencies": {
     "bootstrap-3-typeahead": "^4.0.2",
     "bootstrap-toggle": "^2.2.2",
+    "ci": "^1.0.0",
     "d3": "^3.4.4",
     "d3-tip": "^0.9.1",
-    "dagre-d3": "^0.4.18",
+    "dagre-d3": "^0.6.3",
```
Comment (verdan) [inline airflow/www/package.json:63]: @potiuk did you verify if this is working fine on the frontend?
https://github.com/apache/airflow/pull/4713
https://github.com/apache/airflow/pull/3634

**Review by verdan — CHANGES_REQUESTED** [review]
Just one comment, the rest is looking fine to me.

**Comment by potiuk** [comment]
I would like to still implement it some day :)

**Comment by potiuk** [comment]
I hope I will come back to it after prod image.

**Comment by mik-laj** [comment]
@potiuk I think that we should introduce order in the current code first, and then we can introduce eslint.  We have too much JS code mixed with Jinja, which means that eslint does not work properly. I have already done experiments in this area in the past and were not satisfactory.

**Comment by potiuk** [comment]
Fine. Closing it.

**Pull request closed without merging**